clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -216,16 +216,33 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
 
   // Unary Operators.
   mlir::Value VisitUnaryPostDec(const UnaryOperator *E) {
-    llvm_unreachable("NYI");
+    return buildIncDec(E);
   }
   mlir::Value VisitUnaryPostInc(const UnaryOperator *E) {
-    llvm_unreachable("NYI");
+    return buildIncDec(E);
   }
   mlir::Value VisitUnaryPreDec(const UnaryOperator *E) {
-    llvm_unreachable("NYI");
+    return buildIncDec(E);
   }
   mlir::Value VisitUnaryPreInc(const UnaryOperator *E) {
-    llvm_unreachable("NYI");
+    return buildIncDec(E);
+  }
+  mlir::Value buildIncDec(const UnaryOperator *E) {
```

Comment:
Three uncoded CodeGen best practices:
1. Basically change `Emit` for `build` but maintain the other names. This makes it easier to go back and forth with LLVM codegen to remember or compare approaches. That said, `EmitScalarPrePostIncDec` becomes `buildScalarPrePostIncDec`.
2. Look at the original source and copy the overall layout of the function, but asserting for anything not implemented as part of this, this basically means we only support what we can write tests for. Also same rationale as item 1 + it makes it easier for us to incorporate and catchup with LLVM codegen in case something gets added there. See `buildScalarConversion` vs `EmitScalarConversion` for a concrete example.
3. When you cannot assert on some check because it's not available in CIR codegen, you can add a dummy function in `clang/lib/CIR/CodeGen/UnimplementedFeatureGuarding.h` and assert on that. This way, once we decide to add something we can flip the function result and crash in all places needing support for that.

Reply:
Incorporated the same "skeleton" from `EmitScalarPrePostIncDec`, added asserts accordingly.